### PR TITLE
[Frontend] Add template related params to request

### PR DESCRIPTION
**Pull request by danieljannai21**
HuggingFace's `apply_chat_template` functions allows sending additional parameters to either change the chat template or its behavior. I added `chat_template` and `template_kwargs` to the `ChatCompletionRequest` class to allow full control over the behavior of `apply_chat_template`.
In addition to that, the latest `transformers` version added two explicit parameters that can be sent to `apply_chat_template`: `tools` and `documents`. I added the `documents` field to the `ChatCompletionRequest` class, and converted the existing `tools` field to a list of dicts, prior to passing them both to `apply_chat_template`.

### CODE REVIEW

**Comment by simon-mo** [comment]
Can you add a test case here https://github.com/vllm-project/vllm/blob/main/tests/entrypoints/test_openai_server.py

I'm also curious about the use case for this? In what scenario would you need dynamically supplied chat template instead of a global template for the model (which we already support)?

**Comment by danieljannai21** [comment]
Hi @simon-mo,

Thanks for your answer!

- Regarding the `chat_template` argument: In HF, the `self.chat_template` attribute of the tokenizer can be a dictionary, in which case the `chat_template` argument of the `apply_chat_template` method (yes, I know, bad naming) can be a key that is used to determine which actual template to use (see the block that starts [here](https://github.com/huggingface/transformers/blob/d4564df1d4a6b355779f1a8ac250cb47cb4c38d8/src/transformers/tokenization_utils_base.py#L1770)).
This is used, for example, by Cohere, that have a different template for [tool use](https://github.com/huggingface/transformers/blob/d4564df1d4a6b355779f1a8ac250cb47cb4c38d8/src/transformers/models/cohere/tokenization_cohere_fast.py#L561) and [rag](https://github.com/huggingface/transformers/blob/d4564df1d4a6b355779f1a8ac250cb47cb4c38d8/src/transformers/models/cohere/tokenization_cohere_fast.py#L676).

- Regarding the `chat_template_kwargs` argument: This is used in order pass additional arguments to the template besides `conversation`. Such arguments could be, for example, a list of available tools or reference documents, additional information about the user, a default system message to use, etc. Basically, this can be used for supplying any additional and non-frozen information that doesn't fit in the list of messages (that is the only attribute that's taken from `conversation`.

I hope that makes it more clear 🙂.

File: vllm/entrypoints/openai/protocol.py [hunk 1]
```diff
@@ -190,6 +190,27 @@ class ChatCompletionRequest(OpenAIBaseModel):
...
     )
+    documents: Optional[List[Dict[str, str]]] = Field(
+        default=None,
+        description=
+        ("A list of dicts representing documents that will be accessible to "
+         "the model if it is performing RAG (retrieval-augmented generation)."
+         " If the template does not support RAG, this argument will have no "
+         "effect. We recommend that each document should be a dict containing "
+         "\"title\" and \"text\" keys."),
+    )
```
Comment (simon-mo) [inline vllm/entrypoints/openai/protocol.py:201]: why is this not part of chat_template_kwargs?

**Review by simon-mo — COMMENTED** [review]
ping on the test, thanks!

File: vllm/entrypoints/openai/protocol.py (hunk 1 above)
Comment (danieljannai21) [reply vllm/entrypoints/openai/protocol.py:201]: For consistency with the parameters in HF's [apply_chat_template](https://github.com/huggingface/transformers/blob/3345ae733b6f4aeb7204a0f3e646a3cdbaad0023/src/transformers/tokenization_utils_base.py#L1687)

**Comment by danieljannai21** [comment]
> ping on the test, thanks!

@simon-mo 
Regarding the test -- it's a hard thing to do, as the only thing I'm doing here is passing parameters to external (HF) functions, parameters that are already supported there.
I can't really write a test around `create_chat_completion`, as I can't guarantee anything about the output of a model (since it had to be trained with a template that uses these extra parameters), and writing a test around `apply_chat_template` also feels like a weird thing to do, as its an external function and my only contribution is making sure more parameters (that it supports!) are being passed.

**Review by simon-mo — APPROVED** [review]

**Pull request merged**